File: src/pocketcloud/scheduler/TaskScheduler.php
<?php

namespace pocketcloud\scheduler;

use JetBrains\PhpStorm\Pure;
use pocketcloud\plugin\CloudPlugin;
use pocketcloud\util\Tickable;

class TaskScheduler implements Tickable {

    /** @var array<TaskHandler> */
    private array $tasks = [];

    public function __construct(private CloudPlugin $owner) {}

    private function scheduleTask(Task $task, int $delay, int $period, bool $repeat): void {
        $taskHandler = new TaskHandler($task, $delay, $period, $repeat, $this->owner);
        $task->setTaskHandler($taskHandler);
        $this->tasks[$taskHandler->getId()] = $taskHandler;
    }

    public function scheduleDelayedTask(Task $task, int $delay) {
        $this->scheduleTask($task, $delay, -1, false);
    }

    public function scheduleRepeatingTask(Task $task, int $period) {
        $this->scheduleTask($task, -1, $period, true);
    }

    public function scheduleDelayedRepeatingTask(Task $task, int $delay, int $period) {
        $this->scheduleTask($task, $delay, $period, true);
    }

    public function cancel(Task $task) {
        if (isset($this->tasks[$task->getTaskHandler()->getId()])) {
            $task->getTaskHandler()->cancel();
            unset($this->tasks[$task->getTaskHandler()->getId()]);
        }
    }

    public function cancelAll() {
        foreach ($this->tasks as $task) $task->cancel();
        $this->tasks = [];
    }

    #[Pure] public function getTaskById(int $id): ?Task {
        foreach ($this->tasks as $task) if ($task->getId() == $id) return $task->getTask();
        return null;
    }

    public function tick(int $currentTick): void {
        foreach ($this->tasks as $id => $task) {
            if ($task->isCancelled()) {
                unset($this->tasks[$id]);
                continue;
            }
            $task->onUpdate($currentTick);
        }
    }

    public function getTasks(): array {
        return $this->tasks;
    }

    public function getOwner(): CloudPlugin {
        return $this->owner;
    }
}

File: src/main/java/de/pocketcloud/cloud/scheduler/TaskScheduler.java
package de.pocketcloud.cloud.scheduler;

import de.pocketcloud.cloud.plugin.CloudPlugin;
import de.pocketcloud.cloud.tick.Tickable;
import lombok.Getter;

import java.util.HashMap;
import java.util.Map;
import java.util.Optional;
import java.util.function.Consumer;

public final class TaskScheduler implements Tickable {

    private final Map<Long, TaskHandler> tasks = new HashMap<>();
    @Getter
    private final CloudPlugin owner;

    public TaskScheduler(CloudPlugin owner) {
        this.owner = owner;
    }

    @Override
    public void tick(long currentTick) {
        for (Map.Entry<Long, TaskHandler> entry : tasks.entrySet()) {
            if (entry.getValue().isCancelled()) {
                tasks.remove(entry.getKey());
                continue;
            }

            entry.getValue().update(currentTick);
        }
    }

    public void scheduleDelayedTask(Task task, int delay) {
        scheduleTask(task, delay, 0, false);
    }

    public void scheduleDelayedTask(Consumer<Long> task, int delay) {
        scheduleTask(task, delay, 0, false);
    }

    public void scheduleRepeatingTask(Task task, int interval) {
        scheduleTask(task, 0, interval, true);
    }

    public void scheduleRepeatingTask(Consumer<Long> task, int interval) {
        scheduleTask(task, 0, interval, true);
    }

    public void scheduleDelayedRepeatingTask(Task task, int delay, int interval) {
        scheduleTask(task, delay, interval, true);
    }

    public void scheduleDelayedRepeatingTask(Consumer<Long> task, int delay, int interval) {
        scheduleTask(task, delay, interval, true);
    }

    private void scheduleTask(Task task, int delay, int interval, boolean repeating) {
        TaskHandler taskHandler = new TaskHandler(task, delay, interval, repeating, owner);
        task.setHandler(taskHandler);
        tasks.put(taskHandler.getId(), taskHandler);
    }

    private void scheduleTask(Consumer<Long> task, int delay, int interval, boolean repeating) {
        scheduleTask(task, null, delay, interval, repeating);
    }

    private void scheduleTask(Consumer<Long> task, Consumer<Long> onCancel, int delay, int interval, boolean repeating) {
        scheduleTask(new ClosureTask(task, onCancel),  delay, interval, repeating);
    }

    public void cancel(Task task) {
        if (tasks.containsKey(task.getHandler().getId())) {
            tasks.get(task.getHandler().getId()).cancel();
            tasks.remove(task.getHandler().getId());
        }
    }

    public void cancelAll() {
        for (TaskHandler task : tasks.values()) cancel(task.getTask());
    }

    public Optional<TaskHandler> get(long id) {
        return Optional.ofNullable(tasks.get(id));
    }

    public Map<Long, TaskHandler> getAll() {
        return tasks;
    }
}